File: experiments/phase6/fixtures/python/baseline/gate_helper.py
"""Dependency-free deterministic Gates for the Python Fixture."""

from __future__ import annotations

import ast
import sys
from pathlib import Path

from tag_normalizer import normalize_tags


def _acceptance() -> bool:
    actual = normalize_tags(
        [
            "  Hello World  ",
            "hello__world",
            "Alpha_beta",
            " --Trim-- ",
            "___",
            "ALPHA BETA",
            "x   y",
            "a- _b",
        ]
    )
    return actual == ["hello-world", "alpha-beta", "trim", "x-y", "a--b"]


def _regression() -> bool:
    return normalize_tags([]) == [] and normalize_tags(["plain-tag"]) == [
        "plain-tag"
    ]


def _lint() -> bool:
    content = Path("tag_normalizer.py").read_text(encoding="utf-8")
    lines = content.splitlines()
    return (
        "def normalize_tags(tags: list[str]) -> list[str]:" in content
        and "\t" not in content
        and all(line == line.rstrip() for line in lines)
    )


def _typecheck() -> bool:
    content = Path("tag_normalizer.py").read_text(encoding="utf-8")
    tree = ast.parse(content, filename="tag_normalizer.py")
    functions = [node for node in tree.body if isinstance(node, ast.FunctionDef)]
    return len(functions) == 1 and functions[0].name == "normalize_tags"


GATES = {
    "acceptance": _acceptance,
    "regression": _regression,
    "lint": _lint,
    "typecheck": _typecheck,
}


def main() -> int:
    if len(sys.argv) != 2 or sys.argv[1] not in GATES:
        return 2
    return 0 if GATES[sys.argv[1]]() else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: experiments/phase6/fixtures/typescript/baseline/gate_helper.mjs
/** Dependency-free deterministic Gates for the TypeScript Fixture. */

import { spawnSync } from "node:child_process";
import { mkdirSync, readFileSync } from "node:fs";
import { createRequire } from "node:module";
import { join } from "node:path";

const gate = process.argv[2];
const compilerNode = process.argv[3];
const compilerJs = process.argv[4];
const output = join(process.env.TMPDIR, `phase6-ts-${process.pid}`);
mkdirSync(output, { recursive: true });

function compile(noEmit = false) {
  const args = [
    "--strict",
    "--target",
    "ES2020",
    "--module",
    "commonjs",
    "--skipLibCheck",
  ];
  if (noEmit) {
    args.push("--noEmit");
  } else {
    args.push("--outDir", output);
  }
  args.push("tag_normalizer.ts");
  return spawnSync(compilerNode, [compilerJs, ...args], {
    encoding: "utf8",
    env: process.env,
    shell: false,
  });
}

function loadImplementation() {
  const compiled = compile(false);
  if (compiled.status !== 0) {
    return null;
  }
  const require = createRequire(import.meta.url);
  return require(join(output, "tag_normalizer.js")).normalizeTags;
}

function acceptance() {
  const normalizeTags = loadImplementation();
  if (normalizeTags === null) {
    return false;
  }
  const actual = normalizeTags([
    "  Hello World  ",
    "hello__world",
    "Alpha_beta",
    " --Trim-- ",
    "___",
    "ALPHA BETA",
    "x   y",
    "a- _b",
  ]);
  return JSON.stringify(actual) === JSON.stringify([
    "hello-world",
    "alpha-beta",
    "trim",
    "x-y",
    "a--b",
  ]);
}

function regression() {
  const normalizeTags = loadImplementation();
  return normalizeTags !== null
    && JSON.stringify(normalizeTags([])) === "[]"
    && JSON.stringify(normalizeTags(["plain-tag"])) === '["plain-tag"]';
}

function lint() {
  const content = readFileSync("tag_normalizer.ts", "utf8");
  return content.includes(
    "export function normalizeTags(tags: string[]): string[]",
  ) && !content.includes("\t")
    && content.split("\n").every((line) => line === line.trimEnd());
}

function typecheck() {
  return compile(true).status === 0;
}

const gates = { acceptance, regression, lint, typecheck };
if (!(gate in gates)) {
  process.exit(2);
}
process.exit(gates[gate]() ? 0 : 1);
